### ISSUE TITLE
RFE: access control to resolved for resolving LLMNR/mDNS names, for hookup with flatpak/bubblewrap

### ISSUE THREAD

**Issue by hadess** [systemd/systemd#18860]
With the end goal being to be able to choose whether applications can, or can't do `.local` hostname lookups (and [allowing them to do so in the first place](https://github.com/flatpak/xdg-desktop-portal/issues/505)).

There are a couple of different options that could work:
- ship an older version of `nss_resolve` which uses D-Bus in the sandbox, we could then filter any queries made on the system bus (either using Flatpak's D-Bus proxies, or in systemd-resolved itself)
- the above, but create a D-Bus portal that replicates systemd-resolved's D-Bus interface on the session bus (that's where the authorisation for doing lookups would be queried)
- export `/run/systemd/resolve/io.systemd.Resolve` inside the sandbox, but then systemd-resolved itself would need to check the app for whether it is indeed allowed to access (eg. parsing the `/proc/<client pid>/root/.flatpak-info` keyfile)

If the D-Bus interface is here to stay, we could preempt the availability of `nss_resolve` and implement lookups as necessary in libraries themselves (say, GLib and curl, from the top of my head).

**Comment by poettering** [comment]
what is "sandbox" precisely? do you mean bubblewrap?

**Comment by nanonyme** [comment]
This will start working next Fall when runtimes grow systemd247 support. flatpak started mounting the systemd-resolved varlink socket automatically into sandbox if --share=network is given.

**Comment by nanonyme** [comment]
Using the Dbus is a total no-go. The fallback path is known broken and results in hangs.

**Comment by poettering** [comment]
hmm, what kind of access control do you want precisely? we could offer a separate varlink socket that only allows resolving non-mDNS/LLMNR stuff. Then you could chose between mounting one or the other or neither. Would that work?

**Comment by nanonyme** [comment]
Note the varlink socket is already mounted out of the box with no further access control and this setup works assuming runtime has libsystemd (and glibc NSS plugin) 247 or higher. It's more or less read-only interface. The biggest questionable things I could see is you can query certain security parameters through it and flush DNS cache. Actually modifying configurations is already prevented. (you cannot use policykit nor root from within flatpak)

**Comment by poettering** [comment]
The current varlink socket only allows resolving IP addresses and hostnmes. Nothing else. We can certainly declare that we don#t expose anything else on the socket, and if we want to add more calls such as cache flushing or changing of configuration we do that on another varlink socket independent of it. But even if we didn't do that, we'd check client privileges anyway for those operations, so unless you can be root inside your sandbox we'd deny your requests.

**Comment by nanonyme** [comment]
Right, which you cannot be inside flatpak. https://github.com/flatpak/flatpak/commit/e5da98ff4b12a080044d58b4cd2fb4edb4491132 is the relevant commit

**Comment by hadess** [comment]
> what is "sandbox" precisely? do you mean bubblewrap?

I meant Flatpak (which uses bubblewrap), but I guess this also applies to other types of similar sandboxes.

> This will start working next Fall when runtimes grow systemd247 support. flatpak started mounting the systemd-resolved varlink socket automatically into sandbox if --share=network is given since flatpak 1.10.

This is problematic, as sandboxes would go from not being able to resolve local hostnames, to being able to resolve local hostnames without any advertised changes in permissions, from the user's standpoint.

> The current varlink socket only allows resolving IP addresses and hostnmes. Nothing else. We can certainly declare that we don#t expose anything else on the socket, and if we want to add more calls such as cache flushing or changing of configuration we do that on another varlink socket independent of it. But even if we didn't do that, we'd check client privileges anyway for those operations, so unless you can be root inside your sandbox we'd deny your requests.

Could we add some access control in systemd-resolved so that we can check whether those clients are actually allowed to resolve local hostnames before answering? I don't know what the varlink transport allows, but a check would be fairly trivial to implement given a client PID.

**Comment by nanonyme** [comment]
@hadess what is the attack vector that being able to resolve local hostnames would open? You can anyway contact to local network from within flatpak.

**Comment by hadess** [comment]
> @hadess what is the attack vector that being able to resolve local hostnames would open? You can anyway contact to local network from within flatpak.

For the same reasons Apple does on iOS:
https://github.com/flatpak/flatpak/issues/3054#issuecomment-789810056

We don't have the same sandboxing (we should eventually be able to forbid broadcast networking in Flatpak though), but we're not going to get there by opening the gates wider and wider every time we add features to OS components.

**Comment by poettering** [comment]
> > 
> 
> Could we add some access control in systemd-resolved so that we can check whether those clients are actually allowed to resolve local hostnames before answering? I don't know what the varlink transport allows, but a check would be fairly trivial to implement given a client PID.

If we'd offer you two distinct sockets in the file system, let's say `/run/systemd/resolve/io.systemd.Resolve` and `/run/systemd/resolve/io.systemd.ResolveRemoteOnly` or so, and both implement the very same interfaces, but the latter only allows classic DNS lookups, but nothing of local scope, that should be enough for you, no? 

Having two distinct sockets for this has nice benefits — as long as such a simple two-level policy is sufficient. Because we could set ACLs on the inodes if we wanted to, and you could choose which one of the two you bind mount into your sandbox (you could mount either under the name `/run/systemd/resolve/io.systemd.Resolve`.

Of course, we could also do PID-based access control, but PIDs suck hard (space too small and given how quickly they are recycled they are awful for authentication purposes), we actually try to get away from them as much as we can, and move to better identifiers, such as cgroups or pidfds, depending what works in a specific context.

We can nowadays set xattrs on cgroups. We could use that for encoding per-cgroup policy, i.e. whenever we get a resolution request we could determine its cgroup, and then read an xattr `user.resolve-policy` or so the client is located in. We probably would have to walk recursively up the tree, so that lower cgroups can only take away privs, never add them.

This would then mean, flatpak could do a simple setxattr() on the cgroup path of its sandbox and resolved would honour it.

(The other option would be Polkit, of course, but I think this would suck for this case, since it's so slow requiring so many async roundtrips, and we don't want interactivity for this anyway. I think polkit should work for varlink too if we wanted to)

**Comment by topimiettinen** [comment]
Regarding the goal of choosing whether services can resolve `.local`, maybe the DNS access control feature which I've drafted in #17126 could be useful?

**Comment by nanonyme** [comment]
@hadess right, but you were told that it's not on kernel level currently possible to do the iOS level network distinction with user namespaces. That might change in some distant future or it might not. One thing I noticed that this request is now going in opposite direction to the original RFE from the end-user which was about requesting access to *only* local network, *not* Internet.

**Comment by poettering** [comment]
@nanonyme hmm? not sure what network routing got to do with user namespaces. But the kernel has all the right concepts these days to implement something like the macos thing on linux: for example I think it would make sense to introduce three well-known VRF or so on the system level: one for apps that shall get full network access, one for apps that shall get only remote network access via any network, and one for apps that shall get only remote networks via cell phone only. Then use the cgroup bpf stuff to bind each individual app's traffic to one of these three VRFs (i.e. this is what "ip vrf exec" does internally). That should give you control how stuff is routed, and you can add routes to each that block certain parts of the network.

Alternatively you can us the cgroup bpf stuff directly to influence what processes from specific cgroups can bind on or connect/sendmsg to.

In either case you need your networking solution to maintain a list of subnets that shell be OK (or those which shall not). For the VRF case that should be a routing table, for the other idea it would be a bpf map.

**Comment by mcatanzaro** [comment]
> Using the Dbus is a total no-go. The fallback path is known broken and results in hangs. Or was? I'm not entirely sure fallback was fixed. It definitely hasn't gotten sufficient amount of testing and exposing it would probably not be worth the risk. It's only half a year until we get the varlink interface to work as expected.

That was just a bug, already fixed by https://src.fedoraproject.org/rpms/systemd/c/779685bf4b1cdb74f6f20a6153299178a565e506?branch=f33. That fix never went upstream because the affected code no longer exists upstream. If there is a desire to go back to speaking D-Bus instead of varlink, there is no fundamental problem with doing that.

I don't think it's necessary though. The option of providing two varlink sockets sounds more appealing to me. That way, nss-resolve doesn't require any changes.

> This is problematic, as sandboxes would go from not being able to resolve local hostnames, to being able to resolve local hostnames without any advertised changes in permissions, from the user's standpoint.

Blocking sandboxed apps from resolving .local names sounds fine to me. I never imagined this was something you would want to do when considering the security implications of mounting the varlink socket in the sandbox, but I can see the privacy benefits, especially in a future where users might install untrusted malicious apps, which is what we're designing for. I wouldn't want a hypothetical Home Depot flatpak checking for the existence of particular .local services and using that for fingerprinting purposes, for example.

Unfortunately any real network-level restrictions would require kernel work, since network namespaces are currently all-or-nothing (especially awkward for abstract sockets). But I guess blocking name resolution could at least make it *harder* to do naughty things.

> Regarding the goal of choosing whether services can resolve .local, maybe the DNS access control feature which I've drafted in #17126 could be useful?

Looks like it's based on selinux contexts? selinux is not part of the flatpak security model at all. How would flatpaks and other sandboxed apps (snaps, etc.) be able to take advantage of that?

**Comment by mcatanzaro** [comment]
> Unfortunately any real network-level restrictions would require kernel work, since network namespaces are currently all-or-nothing (especially awkward for abstract sockets). But I guess blocking name resolution could at least make it harder to do naughty things.

Eh, I had never heard of VRFs before Lennart's comment just now. ip-vrf(8) looks interesting. Huh.

**Comment by nanonyme** [comment]
> If we'd offer you two distinct sockets in the file system, let's say `/run/systemd/resolve/io.systemd.Resolve` and `/run/systemd/resolve/io.systemd.ResolveRemoteOnly` or so, and both implement the very same interfaces, but the latter only allows classic DNS lookups, but nothing of local scope, that should be enough for you, no?

This proposal seems straightforward for flatpak to me if the restriction is desired. The resolver plugin could always only use /run/systemd/resolve/io.systemd.Resolve and flatpak could decide which host socket to mount if either.

**Comment by topimiettinen** [comment]
> > Regarding the goal of choosing whether services can resolve .local, maybe the DNS access control feature which I've drafted in #17126 could be useful?
> 
> Looks like it's based on selinux contexts? selinux is not part of the flatpak security model at all. How would flatpaks and other sandboxed apps (snaps, etc.) be able to take advantage of that?

Actually I dropped the SELinux contexts from the later versions. System services can just include a drop-in with line `DNSAllowedDomains=*.local` or `DNSDeniedDomains=*.local`. In order to make this friendly to also user services, minor changes would be needed so that the user unit of individual apps can be detected (instead of all falling to the system unit `user@UID.service`).

**Comment by poettering** [comment]
btw, to give a bit of background: the main reason we switched from dbus to varlink for the nss stuff is that we needed this to work in early boot. And latency is a bit better too, given then varlink requests aren#t bounced of a broker process. both apis will continue to be supported though, it's up to clients to make a choice what they prefer. They can now opt to use one of these four APIs:

1. The varlink one
2. The D-Bus one
3. Talk classic DNS/UDP (or TCP) via 127.0.0.53 stub (i.e. what we write into /etc/resolv.conf)
4. The NSS one (which will either go via nss-resolve → varlink, or nss-dns → DNS/UDP (or TCP))

Each has benefits and drawbacks. (varlink is not so established RPC spec, but works in early boot, brokerless and give you rich info; dbus is well established everywhere, give you rich info, but only late boot. The dns 127.0.0.53 stub is not ideal for mdns/llmnr, and the NSS one is strictly synchronous for something network involved, which sucks, but extremely standard.

Ideally, a sandboxing environment would leave all four ways open in one way or another.

**Comment by poettering** [comment]
> Unfortunately any real network-level restrictions would require kernel work, since network namespaces are currently all-or-nothing (especially awkward for abstract sockets). But I guess blocking name resolution could at least make it _harder_ to do naughty things.

As mentioned above cgroup/bpf stuff already should give you all you need if you want to implement this. 

AF_UNIX abstract namespace sockets are generally a bad idea, people really shouldn't use them.

**Comment by poettering** [comment]
(of course, thinking about this, the fact that we have 4 (well, actually 3) supported APIs to talk to resolved from a local program means that the varlink alternative-socket thing is never going to be enough. Especially the local DNS stub is tough to control, because IP sockets even when used only locally won't tell us anything about the sender. We'd probably have to use VRF or so to make this work, and have a different routing table that route 127.0.0.53 differently for each sandbox

**Comment by hadess** [comment]
> System services can just include a drop-in with line `DNSAllowedDomains=*.local` or `DNSDeniedDomains=*.local`. In order to make this friendly to also user services, minor changes would be needed so that the user unit of individual apps can be detected (instead of all falling to the system unit `user@UID.service`).

If it's systemd-resolved handling the access control, and we can integrate something that'd be Flatpak specific in systemd-resolved, in the same place you're doing those checks, it would certainly be easiest way to fix this issue.

**Comment by nanonyme** [comment]
@poettering note that the D-Bus access is being filtered out of the box anyway. The only stub resolver is an interesting point. I was earlier wondering if flatpak should mount /run/system/resolve/resolv.conf to /etc/resolv.conf if host /etc/resolv.conf is a symlink to /run/system/resolv/stub-resolv.conf to avoid automatically telling everything in sandbox that they're supposed to use stub resolver. I guess there are some drawbacks to that though eg in DNSSEC area.... But yeah, it's somewhat untrivial to prevent someone just running a DNS client in the sandbox from directly connecting to 127.0.0.53 which is well-known IP for stub resolver.
